linmod/eval.py

```diff
@@ -1,124 +1,210 @@
+        and days.
 
-    $\sum_{t, g} E[ || f_{tg} - \phi_{tg} ||_p ]$
-    """
+        $\sum_{t, g} E[ || f_{tg} - \phi_{tg} ||_p ]$
+        """
 
-    return (
-        proportions_mean_norm_per_division_day(samples, data, p=p)
-        .collect()
-        .get_column("mean_norm")
-        .sum()
-    )
+        return (
+            self._mean_norm_per_division_day(p=p)
+            .collect()
+            .get_column("mean_norm")
+            .sum()
+        )
 
+    def _energy_score_per_division_day(self, p=2) -> pl.LazyFrame:
+        r"""
+        Monte Carlo approximation to the energy score (multivariate generalization of
+        CRPS) of proportion forecasts for each division-day.
+
+        $E[ || f_{tg} - \phi_{tg} ||_p ] - \frac{1}{2} E[ || f_{tg} - \f_{tg}' ||_p ]$
+
+        Returns a data frame with columns `(division, fd_offset, energy_score)`.
+        """
+
+        return (
+            # First, we will gather the values of phi' we will use for (phi-phi')
+            self.df.group_by("date", "fd_offset", "division", "lineage")
+            .agg(pl.col("sample_index"), pl.col("phi"), pl.col("phi_sampled"))
+            .with_columns(replicate=pl_list_cycle(pl.col("phi_sampled"), 1))
+            .explode("sample_index", "phi", "phi_sampled", "replicate")
+            # Now we can compute the score
+            .group_by("fd_offset", "division", "sample_index")
+            .agg(
+                term1=pl_norm(pl.col("phi") - pl.col("phi_sampled"), p),
+                term2=pl_norm(
+                    (pl.col("phi_sampled") - pl.col("replicate")), p
+                ),
+            )
+            .group_by("fd_offset", "division")
+            .agg(
+                energy_score=pl.col("term1").mean()
+                - 0.5 * pl.col("term2").mean()
+            )
+        )
 
-def proportions_energy_score_per_division_day(samples, data, p=2):
-    r"""
-    Monte Carlo approximation to the energy score (multivariate generalization of CRPS)
-    of proportion forecasts for each division-day.
+    def energy_score(self, p=2) -> float:
+        r"""
+        The energy score of proportion forecasts, summed over all divisions and days.
+
+        $$
+        \sum_{t, g} E[ || f_{tg} - \phi_{tg} ||_p ]
+        - \frac{1}{2} E[ || f_{tg} - f_{tg}' ||_p ]
+        $$
+        """
+
+        return (
+            self._energy_score_per_division_day(p=p)
+            .collect()
+            .get_column("energy_score")
+            .sum()
+        )
 
-    $E[ || f_{tg} - \phi_{tg} ||_p ] - \frac{1}{2} E[ || f_{tg} - \f_{tg}' ||_p ]$
 
-    `samples` should have the standard model output format.
-    `data` should have the standard model input format.
+class CountsEvaluator:
+    _count_samplers = {
+        "multinomial": multinomial_count_sampler,
+    }
+
+    def __init__(
+        self,
+        samples: pl.LazyFrame,
+        data: pl.LazyFrame,
+        count_sampler: str = "multinomial",
+        seed: int = 42,
```

Comment:
I think I wrote this hard-coded default seed in #50, but I think it's perhaps not a great idea.

We should probably follow NumPyro style here and demand one get passed in.

Reply:
I agree to remove the default seed, but I'm more inclined to follow numpy's idea than numpyro's, which is that "no seed" is a valid configuration, meaning "give me fresh randomness". I propose to set `None` to replace `42` here